manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -267,23 +267,21 @@ func (tr *TaskReaper) tick() {
 
 			runningTasks := 0
 			for _, t := range historicTasks {
-				// Skip tasks which are desired to be running but the current state
-				// is less than or equal to running.
-				// This check is important to ignore tasks which are running or need to be running,
-				// but to delete tasks which are either past running,
-				// or have not reached running but need to be shutdown (because of a service update, for example).
-				if t.DesiredState == api.TaskStateRunning && t.Status.State <= api.TaskStateRunning {
-					// Don't delete running tasks
-					runningTasks++
-					continue
-				}
-
-				deleteTasks[t.ID] = struct{}{}
+				// Historical tasks can be considered for cleanup if:
+				// 1. The task has reached a terminal state i.e. actual state beyond TaskStateRunning.
+				// 2. The task has not yet become running and desired state is a terminal state i.e.
+				// actual state not yet TaskStateAssigned and desired state beyond TaskStateRunning.
+				if t.Status.State > api.TaskStateRunning ||
```

Comment:
This is becoming hard to grasp (also a reason we now need a long comment to explain)

Perhaps we should split these up (may give some duplicated code) or extract  the check to a function; e.g `okToCleanup(task) bool`

Inside that function we can do an early return for each check to reduce complexity

Reply:
ahh fair point. I'll try to make it easier to read.

Reply:
Yes, the comment itself is useful, sorry, didn't meant to imply that, just that it's easy to make mistakes in these conditions; breaking them up makes the code easier to read, and less likely to make mistakes